src/oas.md

```diff
@@ -557,6 +557,23 @@ servers:
 
 An object representing a Server Variable for server URL template substitution.
 
```

Comment:
Hi @baywet,

As we've exchanged in https://github.com/OAI/OpenAPI-Specification/issues/3256#issuecomment-2514431402, IMHO IServer URL templating is distinct thing from Path templating. 

I would suggest following ABNF (It's basically a reduced and modified version of [openapi-server-url-templating](https://github.com/swaggerexpert/openapi-server-url-templating); modifications come from concensus in https://github.com/OAI/OpenAPI-Specification/pull/4244):

```abnf
server-url-template    = 1*( literals / server-variable )
server-variable        = "{" server-variable-name "}"
server-variable-name   = 1*( %x00-79 / %x7C / %x7E-10FFFF ) ; every UTF8 character except { and }
literals               = 1*( %x21 / %x23-24 / %x26 / %x28-3B / %x3D / %x3F-5B
                       / %x5D / %x5F / %x61-7A / %x7E / ucschar / iprivate
                       / pct-encoded)
                           ; any Unicode character except: CTL, SP,
                           ;  DQUOTE, "'", "%" (aside from pct-encoded),
                           ;  "<", ">", "\", "^", "`", "{", "|", "}"

pct-encoded    =  "%" HEXDIG HEXDIG
ucschar        =  %xA0-D7FF / %xF900-FDCF / %xFDF0-FFEF
               /  %x10000-1FFFD / %x20000-2FFFD / %x30000-3FFFD
               /  %x40000-4FFFD / %x50000-5FFFD / %x60000-6FFFD
               /  %x70000-7FFFD / %x80000-8FFFD / %x90000-9FFFD
               /  %xA0000-AFFFD / %xB0000-BFFFD / %xC0000-CFFFD
               /  %xD0000-DFFFD / %xE1000-EFFFD
iprivate       =  %xE000-F8FF / %xF0000-FFFFD / %x100000-10FFFD
```

**Reasoning**:

1. Alignment with https://www.rfc-editor.org/rfc/rfc6570. IMHO Server URL templating has nothing to do with RFC 3986, as server variables can appear in any place within the server URL not just the [Path](https://datatracker.ietf.org/doc/html/rfc3986#section-3.3) section as in Path Templating. This means that `literals` non-terminal needs to be used from RFC 6570 to take that into account.
2. Naming - Server URL templates don't don't mention explicitly concept of `template expression` at all. Server URL templating mentions concept of `server variable substitution`. So the ABNF naming is revolving around this. 

---

`literals`, `iprivate`, `ucschar` and `pct-encoded` are taken directly from https://www.rfc-editor.org/rfc/rfc6570

Reply:
Thanks for the feedback!

1. I'm not sure why RFC6570 is relevant here? OAS does not depend on it, it only has a few interop considerations in the appendix.
2. Names: I tried to align with what we've built so far for path templates, but I'm happy to change the names if we want to clear confusion/think they are semantically different.

Reply:
> I'm not sure why RFC6570 is relevant here? OAS does not depend on it, it only has a few interop considerations in the appendix.

I'll try to elaborate fully on this, so that we build shared understanding. I'll explain my reasoning which should help us move forward or dismiss my logic as nonsensical.

The current `server-url-template` non-terminal as defined in this PR, will not parse an absolute URI (https://datatracker.ietf.org/doc/html/rfc3986#section-4.3) like `https://example.com`. It will fail on a server URL template like this as well `{scheme}://example.com`. If you build ABNF grammar on top of RFC 3986 for `server URL templating`, which I've tried during the first implementation of [openapi-server-url-templating](https://github.com/swaggerexpert/openapi-server-url-templating), you'll end up with extremely large and unreliable grammar, which will have to alternate a lot (I've never finished it as it was so ambiguous that I was having problem keeping it together). And then you'll get into a situation when `{server-variable}` can span multiple non-terminals of absolute URI like `scheme` or  `hier-part` (fully or partially). 

`pchar` from `path templating` is used to represent group of literals allowed in [path](https://datatracker.ietf.org/doc/html/rfc3986#section-3.3). It also represents characters allowed in [query](https://datatracker.ietf.org/doc/html/rfc3986#section-3.4) and [fragment](https://datatracker.ietf.org/doc/html/rfc3986#section-3.5) - but as established, query and fragment cannot be part of path template. We specifically use it to express `path-absolute` non-terminal (https://datatracker.ietf.org/doc/html/rfc3986#section-3.3) with support of template expressions. It doesn't encompass allowed characters from the entire URI structure. 

And that is where `path templating` differs from `server URL templating`. `Path templating` only deals with `path-absolute` non-terminal (https://datatracker.ietf.org/doc/html/rfc3986#section-3.3), apart from `server URL templating` which deals with the whole URI. This has signification implications on the ABNF grammar.

Now to avoid all the complexities of alteration and ambiguity described in the second paragraph of this elaboration, we can build on ideas of smart people that defined [RFC 6570](https://www.rfc-editor.org/rfc/rfc6570). They faced exactly the same problem - coming up with ABNF grammar that will allow substituting template expressions for values in any part of the URI. Their ABNF start rule is:

```abnf
URI-Template  = *( literals / expression )
```


So the entire URI si decomposed to either URI literals or template expressions. URI literals must be defined carefully to allow to encompass any part of the URI.

```abnf
 literals      =  %x21 / %x23-24 / %x26 / %x28-3B / %x3D / %x3F-5B
                   /  %x5D / %x5F / %x61-7A / %x7E / ucschar / iprivate
                   /  pct-encoded
                        ; any Unicode character except: CTL, SP,
                        ;  DQUOTE, "'", "%" (aside from pct-encoded),
                        ;  "<", ">", "\", "^", "`", "{", "|", "}"
```

literals in RFC 6570 is like an abstraction layer designed to simplify handling URI templates without having to explicitly parse or manage the individual parts of a URI as defined in RFC 3986.

IMHO We cannot really say that `server URL templating` is a subset of  URI-templating because of following definitions (but IMHO we can say it's derived from it):

```anbf
server-url-template    = 1*( literals / server-variable ) ; vs URI-Template  = *( literals / expression )
server-variable-name   = 1*( %x00-79 / %x7C / %x7E-10FFFF ) ; every UTF8 character except { and }; vs variable-list =  varspec *( "," varspec )
```

The first one is the implication of following OpenAPI excerpt:
> If the servers field is not provided, or is an empty array, the default value would be a [Server Object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.4.md#server-object) with a [url](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.4.md#server-url) value of /

The second one is local consensus of how the server variable name can be defined.

Reply:
I am not sold yet on basing server URL templates on a different RFC than path templates. The main difference between the two is that server URL templates may also contain scheme and authority components in RFC3986 terminology, so I would actually prefer basing both flavors of URL templates on RFC3986.

Reply:
Thank you for taking the time to provide a detailed explanation of your thinking process here.

That's a question I had for the group: do we want to ensure people can't build server URIs that end up being invalid based on the "template" structure?

I was thinking that no we don't want to because:
- it might be more restrictive than what can currently be done/is supported by the current tools. And the work on path templating clearly established we don't want to restrain the "existing situation" any further.
- ultimately, we don't control what's in the variable values, which could render the whole resulting server URI invalid anyway.

This is why I put together this initial draft with simpler grammar that's similar to the path templating, granted we can adjust it to account for missing characters etc...

Reply:
Further discussions on the TSC call today with @ralfhandl and @handrews allowed me to understand things a bit better.
I've just pushed a change with pchar being replaced by literal (and the other definitions you've provided)

I have kept the naming for template-expression and template-expression-param-name, since they are identical to the ones in path templating. We could imagine "refactoring" them, but it'd probably make things harder. At least implementer know they are the same (should we add a comment in that sense?)

Reply:
> I am not sold yet on basing server URL templates on a different RFC than path templates. The main difference between the two is that server URL templates may also contain scheme and authority components in RFC3986 terminology, so I would actually prefer basing both flavors of URL templates on RFC3986.

@ralfhandl, sure I understand the need to stay within the RFC 3986. My initial experimentation about six months about around creating ABNF for Server URL templating in [openapi-server-url-templating](https://github.com/swaggerexpert/openapi-server-url-templating) was actually trying to achieve this. After couple of unsuccessful attempts I ended up with RFC 6570.

> That's a question I had for the group: do we want to ensure people can't build server URIs that end up being invalid based on the "template" structure?

IMHO we MUST NOT. This would break a lot of stuff. We can only say that after successful server variable substitution the server URL must be valid URL or relative reference. So I totally agree with your two conclusions.

> Further discussions on the TSC call today with @ralfhandl and @handrews allowed me to understand things a bit better. I've just pushed a change with pchar being replaced by literal (and the other definitions you've provided).

ACK